src/Microsoft.AspNetCore.Mvc.Core/Formatters/TextOutputFormatter.cs

```diff
@@ -0,0 +1,246 @@
+        /// for reading or writing an HTTP entity body based on the provided <paramref name="contentTypeHeader"/>.
+        /// </summary>
+        /// <param name="context">The formatter context associated with the call.
+        /// </param>
+        /// <returns>The <see cref="Encoding"/> to use when reading the request or writing the response.</returns>
+        public virtual Encoding SelectCharacterEncoding(OutputFormatterWriteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var request = context.HttpContext.Request;
+            var encoding = MatchAcceptCharacterEncoding(request.GetTypedHeaders().AcceptCharset);
+            if (encoding != null)
+            {
+                return encoding;
+            }
+
+            if (context.ContentType.HasValue)
+            {
+                var parsedContentType = new MediaType(context.ContentType);
+                var contentTypeCharset = parsedContentType.Charset;
+                if (contentTypeCharset.HasValue)
+                {
+                    for (var i = 0; i < SupportedEncodings.Count; i++)
+                    {
+                        var supportedEncoding = SupportedEncodings[i];
+                        if (contentTypeCharset.Equals(supportedEncoding.WebName, StringComparison.OrdinalIgnoreCase))
+                        {
+                            // This is supported.
+                            return SupportedEncodings[i];
+                        }
+                    }
+                }
+            }
+
+            return SupportedEncodings.Count > 0 ? SupportedEncodings[0] : null;
+        }
+
+        /// <inheritdoc />
+        public override Task WriteAsync(OutputFormatterWriteContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var selectedMediaType = context.ContentType;
+            if (!selectedMediaType.HasValue)
+            {
+                // If content type is not set then set it based on supported media types.
+                if (SupportedEncodings.Count > 0)
+                {
+                    selectedMediaType = new StringSegment(SupportedMediaTypes[0]);
+                }
+                else
+                {
+                    throw new InvalidOperationException(Resources.FormatOutputFormatterNoMediaType(GetType().FullName));
+                }
+            }
+
+            var selectedEncoding = SelectCharacterEncoding(context);
+            if (selectedEncoding != null)
+            {
+                // Override the content type value even if one already existed.
+                var mediaTypeWithCharset = GetMediaTypeWithCharset(selectedMediaType.Value, selectedEncoding);
+                selectedMediaType = new StringSegment(mediaTypeWithCharset);
+            }
+
```

Comment:
I've implemented the check for SupportedEncodings > 0 inside SelectCharacterEncoding. By default we will never return null from that method as we will return the first character encoding if we can't find an encoding to use from the accept-charset header or from the content type header of the response.

If a user wants to change this behavior, they can override SelectCharacterEncoding, we will continue to work properly and they should take into account null encoding values inside their WriteBodyAsync(OutputFormatterContext context, Encoding encoding) implementation